src/utils/spanUtils.ts

```diff
@@ -0,0 +1,156 @@
+import { URL } from 'url';
+import { ReadableSpan } from '@opentelemetry/tracing';
+import { hrTimeToMilliseconds } from '@opentelemetry/core';
+import { SpanKind, Logger, CanonicalCode, Link } from '@opentelemetry/api';
+import { Envelope, RequestData, RemoteDependencyData, Base } from '../Declarations/Contracts';
+import { Tags, Properties, MSLink } from '../types';
+
+function createTagsFromSpan(span: ReadableSpan): Tags {
+  const tags: Tags = {};
+  tags['ai.operation.id'] = span.spanContext.traceId;
+  if (span.parentSpanId) {
+    tags['ai.operation.parentId'] = span.parentSpanId;
+  }
+  // @todo: is this for RequestData only?
+  // @todo: parse grpc attributes where appropriate
+  if (
+    (span.kind === SpanKind.SERVER || span.kind === SpanKind.CONSUMER) &&
+    span.attributes['http.method'] &&
+    span.attributes['http.route']
+  ) {
+    tags['ai.operation.name'] = `${span.attributes['http.method']} ${span.attributes['http.route']}`;
+  }
+  return tags;
+}
+
+function createPropertiesFromSpan(span: ReadableSpan): Properties {
+  const properties: Properties = {};
+
+  Object.keys(span.attributes).forEach((key: string) => {
+    if (!(key.startsWith('http.') || key.startsWith('grpc.'))) {
+      properties[key] = span.attributes[key] as string;
+    }
+  });
+
+  const links: MSLink[] = span.links.map((link: Link) => ({
+    operation_Id: link.context.traceId,
+    id: link.context.spanId,
+  }));
+
+  properties['_MS.links'] = links;
+
+  return properties;
+}
+
+function createDependencyData(span: ReadableSpan): RemoteDependencyData {
+  const data = new RemoteDependencyData();
+  data.name = span.name;
+  data.id = `|${span.spanContext.traceId}.${span.spanContext.spanId}.`;
```

Comment:
Is the last "." added on purpose? curious what does it concat with?

Reply:
It is required for parentId to get picked up properly